src/main/java/org/candlepin/subscriptions/tally/InventorySwatchDataCollator.java

```diff
@@ -157,6 +157,26 @@ will be processed (and the other will be saved for the next iteration).
       if (activeSwatchSystem != null) {
         // consume the swatch system from the stream to prepare for the next iteration
         swatchDataIterator.next();
+      } else if (activeHbiSystem != null
+          && (activeHbiSystem.getInventoryId() != null
+              || activeHbiSystem.getProviderId() != null)) {
+        // we need to be certain that the swatch system does not exist when the inventory id and the
+        // instance id is provided to prevent unique constraints.
+        // note that this is a performance killer since we are going to perform a query when the
+        // swatch host does not exist yet,
+        // but it's necessary since we might receive duplicated hosts from HBI and if we don't do it
+        // and let the unique constraint to happen, it will rollback all the hosts (duplicated or
+        // not) from the current process.
+        // This won't be necessary after SWATCH-512 is done.
+        activeSwatchSystem =
+            hostRepository
+                .findByOrgIdAndInstanceIdOrInventoryId(
+                    orgId,
+                    activeHbiSystem.getProviderId(),
+                    Optional.ofNullable(activeHbiSystem.getInventoryId())
+                        .map(Object::toString)
+                        .orElse(null))
+                .orElse(null);
```

Comment:
Rather than doing this query for each host, we should rework the query that produces the stream of swatch systems to fetch these systems as well.

Reply:
We can't do this using the stream of swatch systems because the duplicated host might come from the new hosts from the stream HBI query. 

Let me try to explain this better:

- hosts from HBI

hostA
hostB
hostB (host with the same instance ID or inventory ID)

- hosts from SWATCH

hostA

When processing the reconciliation:
- the hostA will be properly updated
- the first hostB will be properly added
- the second hostB will throw an unique constraint because it was added right in the previous iteration

As a result of the unique constraint, neither the hostA nor the hostB will be added/updated.

Reply:
@kahowell PR updated by using "desc nulls last" instead.

Reply:
resolved offline